core/deployment/src/main/java/io/quarkus/deployment/steps/RegisterForReflectionBuildStep.java

```diff
@@ -45,49 +64,159 @@ public void build(CombinedIndexBuildItem combinedIndexBuildItem,
             }
         } catch (ClassNotFoundException e) {
             log.warnf(e, "Failed to load Class %s", className);
         }
     }
 
+    private void registerClassDependencies(BuildProducer<ReflectiveHierarchyBuildItem> reflectiveClassHierarchy,
+            ClassLoader classLoader, Set<DotName> processedReflectiveHierarchies, boolean methods,
+            ReflectiveHierarchyBuildItem.Builder builder,
+            String className) {
+        try {
+            DotName dotName = DotName.createSimple(className);
+            if (!processedReflectiveHierarchies.contains(dotName)
+                    && !ReflectiveHierarchyBuildItem.DefaultIgnoreTypePredicate.INSTANCE.test(dotName)) {
+
+                processedReflectiveHierarchies.add(dotName);
+                Index indexView = Index.of(classLoader.loadClass(className));
+                reflectiveClassHierarchy.produce(builder
+                        .type(org.jboss.jandex.Type
+                                .create(dotName,
+                                        org.jboss.jandex.Type.Kind.CLASS))
+                        .index(indexView)
+                        .build());
+
+                ClassInfo classInfo = indexView.getClassByName(dotName);
+                if (methods) {
+                    addMethodsForSerialization(reflectiveClassHierarchy, classLoader, processedReflectiveHierarchies,
+                            indexView, dotName,
+                            builder, classInfo, methods);
+                }
+                registerClassFields(reflectiveClassHierarchy, classLoader, processedReflectiveHierarchies,
+                        indexView, dotName, builder, classInfo, methods);
+            }
+        } catch (ClassNotFoundException | IOException ignored) {
+            log.error("Cannot load class for reflection: " + className);
+        }
+    }
+
+    private void addMethodsForSerialization(BuildProducer<ReflectiveHierarchyBuildItem> reflectiveClassHierarchy,
+            ClassLoader classLoader, Set<DotName> processedReflectiveHierarchies, IndexView indexView, DotName initialName,
+            ReflectiveHierarchyBuildItem.Builder builder, ClassInfo classInfo, boolean methods) {
+        List<MethodInfo> methodList = classInfo.methods();
+        for (MethodInfo methodInfo : methodList) {
+            // we will only consider potential getters
+            if (methodInfo.parameters().size() > 0 ||
+                    Modifier.isStatic(methodInfo.flags()) ||
+                    methodInfo.returnType().kind() == Kind.VOID || methodInfo.returnType().kind() == Kind.PRIMITIVE) {
+                continue;
+            }
+            registerType(reflectiveClassHierarchy, classLoader, processedReflectiveHierarchies,
+                    methods, builder,
+                    getMethodReturnType(indexView, initialName, classInfo, methodInfo));
+        }
+    }
```

Comment:
Non-blocker: I still don't see why serialization/deserialization would use the getters instead of reflection to access the fields. If you could provide an example where getters are accessed/used I would appreciate it :)

Reply:
Sure. We are not using the getters per se, but only to guess what field should we register the type from. That way we are registering types returned by public methods, since I'm guessing that its state is needed to be transferred or stored.  

If I stop doing that, the opentelemetry vertex integration test break due to lack of type when you run it on native.

Reply:
Why not register the types of all the fields though? Wouldn't that be safer?

Reply:
That crossed my mind, but I've noticed that many types returned on opentelemetry vertex integration test are just interfaces, so they have no fields to be registered and we have no access (the developer may not know and not asking them to be registered) to the real objects behind those interfaces.